File: snapshot/src/fsm_skating/services/engine.py
import os
import yaml
import random
from typing import List, Dict, Tuple, Any, Optional
from pydantic import BaseModel

# 引入重构后的领域模型命名空间
from fsm_skating.domain.models import State, get_relative_conditions, ALL_STATES


def check_match(
    current_state: State, target_state: State, move_config: Dict[str, Any]
) -> bool:
    """
    双重规则匹配逻辑。
    """
    conditions = move_config.get("conditions", {})
    actual_conditions = get_relative_conditions(current_state, target_state)

    if (
        conditions.get("same_foot") != actual_conditions["same_foot"]
        or conditions.get("same_dir") != actual_conditions["same_dir"]
        or conditions.get("same_edge") != actual_conditions["same_edge"]
    ):
        return False

    if "start_constraints" in move_config:
        constraints = move_config["start_constraints"]
        if "dir" in constraints and current_state.direction != constraints["dir"]:
            return False
        if "edge" in constraints and current_state.edge != constraints["edge"]:
            return False

    return True


class Move(BaseModel):
    id: str
    name: str
    category: str
    difficulty: int
    turn_rotation: Optional[str] = None
    conditions: Dict[str, bool]
    start_constraints: Optional[Dict[str, str]] = None
    rotation_dir: Optional[str] = None
    geometry_config: Optional[Dict[str, float]] = None  # 新增透传因子字段


class MoveOption(BaseModel):
    target_state: State
    move: Move


class TransitionDetail(BaseModel):
    from_state: State
    to_state: State
    candidate_moves: List[Move]
    selected_move: Move


class VerificationResponse(BaseModel):
    valid: bool
    error: Optional[str] = None
    states: Optional[List[State]] = None
    transitions: Optional[List[TransitionDetail]] = None
    total_difficulty: int = 0
    is_ambiguous: bool = False


class MoveVerificationDetail(BaseModel):
    from_state: State
    move: Move
    to_state: State


class MoveVerificationResponse(BaseModel):
    valid: bool
    error: Optional[str] = None
    trace: Optional[List[MoveVerificationDetail]] = None
    total_difficulty: int = 0


class ChoreographyEngine:
    """
    花样滑冰状态机编排与过滤引擎。
    """

    def __init__(self, config_path: str):
        self.config_path = config_path
        self.moves = self._load_config()

    def _load_config(self) -> List[Dict[str, Any]]:
        if not os.path.exists(self.config_path):
            raise FileNotFoundError(
                f"Configuration file not found at: {self.config_path}"
            )
        with open(self.config_path, "r", encoding="utf-8") as f:
            data = yaml.safe_load(f)
        return data.get("moves", [])

    def _build_move(self, move_data: Dict[str, Any], current_state: State) -> Move:
        from fsm_skating.domain.models import get_natural_curvature

        turn_rot = move_data.get("turn_rotation")
        abs_rot = None
        if turn_rot == "natural":
            abs_rot = get_natural_curvature(current_state)
        elif turn_rot == "opposite":
            start_curv = get_natural_curvature(current_state)
            abs_rot = "CW" if start_curv == "CCW" else "CCW"

        return Move(
            id=move_data["id"],
            name=move_data["name"],
            category=move_data["category"],
            difficulty=move_data["difficulty"],
            turn_rotation=move_data.get("turn_rotation"),
            conditions=move_data["conditions"],
            start_constraints=move_data.get("start_constraints"),
            rotation_dir=abs_rot,
            geometry_config=move_data.get("geometry_config"),
        )

    def get_possible_transitions(
        self, current_state: State, max_difficulty: int = 999
    ) -> List[MoveOption]:
        """
        管道式过滤核心逻辑：获取当前滑行状态下合规的下一个转移分支。
        """
        results: List[MoveOption] = []

        for target_state in ALL_STATES:
            if target_state == current_state:
                continue

            for move_data in self.moves:
                if check_match(current_state, target_state, move_data):
                    diff = move_data.get("difficulty", 0)
                    if diff <= max_difficulty:
                        move_obj = self._build_move(move_data, current_state)

                        results.append(
                            MoveOption(target_state=target_state, move=move_obj)
                        )

        results.sort(key=lambda x: (x.move.difficulty, x.move.name))
        return results

    def verify_sequence(self, sequence_str: str) -> VerificationResponse:
        """
        验证用刃状态序列，解析出每一个物理动作及难度评分。
        """
        parts = [p.strip().upper() for p in sequence_str.split("->") if p.strip()]
        states: List[State] = []
        for part in parts:
            try:
                states.append(State.from_string(part))
            except ValueError as e:
                return VerificationResponse(
                    valid=False,
                    error=f"状态字符 '{part}' 格式有误: {str(e)}",
                )

        if len(states) < 2:
            return VerificationResponse(
                valid=False,
                error="状态序列中至少需要包含 2 个有效状态才能进行转移校验。",
            )

        transitions_details: List[TransitionDetail] = []
        total_difficulty = 0

        for i in range(len(states) - 1):
            s_from = states[i]
            s_to = states[i + 1]

            if s_from == s_to:
                return VerificationResponse(
                    valid=False,
                    error=f"第 {i + 1} 步转移出现原地停滞 ({s_from} -> {s_to})，这不符合动力学步法转移规则。",
                )

            matched_moves: List[Move] = []
            for move_data in self.moves:
                if check_match(s_from, s_to, move_data):
                    move_obj = self._build_move(move_data, s_from)
                    matched_moves.append(move_obj)

            if not matched_moves:
                return VerificationResponse(
                    valid=False,
                    error=f"无法识别的物理转移: 从状态 {s_from} 无法直接通过任何已知动作转移到 {s_to}。",
                )

            matched_moves.sort(key=lambda m: (m.difficulty, m.name))

            transitions_details.append(
                TransitionDetail(
                    from_state=s_from,
                    to_state=s_to,
                    candidate_moves=matched_moves,
                    selected_move=matched_moves[0],
                )
            )
            total_difficulty += matched_moves[0].difficulty

        is_ambiguous = any(len(t.candidate_moves) > 1 for t in transitions_details)

        return VerificationResponse(
            valid=True,
            states=states,
            transitions=transitions_details,
            total_difficulty=total_difficulty,
            is_ambiguous=is_ambiguous,
        )

    def verify_move_sequence(
        self, move_ids: List[str], start_state: Optional[State] = None
    ) -> MoveVerificationResponse:
        """
        动作驱动校验器：输入动作 ID 序列，自动演化滑行状态并进行全段物理合法性校验。
        """
        if not move_ids:
            return MoveVerificationResponse(valid=False, error="动作 ID 序列不能为空。")

        # 缓存配置字典，便于 O(1) 检索
        move_db = {m["id"]: m for m in self.moves}

        # 确定起始滑行状态
        current_state = start_state
        if not current_state:
            # 尝试推导首个动作的缺省起滑状态
            first_move_data = move_db.get(move_ids[0])
            if not first_move_data:
                return MoveVerificationResponse(
                    valid=False, error=f"无法识别序列起始处的动作 ID: {move_ids[0]}"
                )
            constraints = first_move_data.get("start_constraints", {})
            dir_c = constraints.get("dir", "F")
            edge_c = constraints.get("edge", "O")
            # 缺省选用左脚（L），完全符合标准的编排惯例
            current_state = State(foot="L", direction=dir_c, edge=edge_c)

        trace_details: List[MoveVerificationDetail] = []
        total_difficulty = 0

        for idx, move_id in enumerate(move_ids):
            move_data = move_db.get(move_id)
            if not move_data:
                return MoveVerificationResponse(
                    valid=False,
                    error=f"在第 {idx + 1} 步检测到未知动作 ID: '{move_id}'",
                )

            # 1. 起滑方向与用刃约束检查
            constraints = move_data.get("start_constraints")
            if constraints:
                if (
                    "dir" in constraints
                    and current_state.direction != constraints["dir"]
                ):
                    return MoveVerificationResponse(
                        valid=False,
                        error=f"第 {idx + 1} 步动作校验失败：动作 '{move_data['name']}' 要求以 '{constraints['dir']}' 向起滑，但当前滑行状态为 '{current_state}'。",
                    )
                if "edge" in constraints and current_state.edge != constraints["edge"]:
                    return MoveVerificationResponse(
                        valid=False,
                        error=f"第 {idx + 1} 步动作校验失败：动作 '{move_data['name']}' 要求以 '{constraints['edge']}' 内外刃起滑，但当前滑行状态为 '{current_state}'。",
                    )

            # 2. 调用第一阶段新引入的状态推导核心演算下一个状态
            from fsm_skating.domain.models import (
                calculate_next_state,
            )

            conditions = move_data["conditions"]
            next_state = calculate_next_state(current_state, conditions)

            # 3. 统一推导旋转体并构建 Move 对象
            move_obj = self._build_move(move_data, current_state)

            trace_details.append(
                MoveVerificationDetail(
                    from_state=current_state,
                    move=move_obj,
                    to_state=next_state,
                )
            )
            total_difficulty += move_obj.difficulty
            current_state = next_state

        return MoveVerificationResponse(
            valid=True,
            trace=trace_details,
            total_difficulty=total_difficulty,
        )

    def generate_sequence(
        self, steps: int, max_difficulty: int, start_state: Optional[State] = None
    ) -> Optional[List[Tuple[State, Optional[Move]]]]:
        """
        智能随机生成模块。
        """
        if steps <= 0:
            return []

        init_state = start_state if start_state else random.choice(ALL_STATES)

        def dfs(
            curr_state: State, remaining_steps: int
        ) -> Optional[List[Tuple[State, Move]]]:
            if remaining_steps == 0:
                return []

            options = self.get_possible_transitions(curr_state, max_difficulty)
            if not options:
                return None

            shuffled_options = list(options)
            random.shuffle(shuffled_options)

            for opt in shuffled_options:
                nxt_state = opt.target_state
                move = opt.move

                sub_path = dfs(nxt_state, remaining_steps - 1)
                if sub_path is not None:
                    return [(nxt_state, move)] + sub_path

            return None

        sub_path = dfs(init_state, steps)
        if sub_path is None:
            return None

        result: List[Tuple[State, Optional[Move]]] = []
        curr = init_state
        for nxt_state, move in sub_path:
            result.append((curr, move))
            curr = nxt_state
        result.append((curr, None))

        return result

    def check_library_integrity(self) -> Dict[str, Any]:
        """
        检查动作库中各类别动作的覆盖度 (FO, FI, BO, BI)。
        """
        with open(self.config_path, "r", encoding="utf-8") as f:
            data = yaml.safe_load(f)

        categories = data.get(
            "categories",
            {
                "three_turn": "转三步 (Three-Turn)",
                "bracket": "括弧步 (Bracket)",
                "rocker": "摇滚步 (Rocker)",
                "counter": "计数步 (Counter)",
                "mohawk": "莫霍克步 (Mohawk)",
                "choctaw": "乔克陶步 (Choctaw)",
            },
        )

        required = ["FO", "FI", "BO", "BI"]
        report = {}

        for cat_id, cat_name in categories.items():
            report[cat_id] = {
                "name": cat_name,
                "implemented": [],
                "missing": list(required),
                "generic_count": 0,
            }

        for move in self.moves:
            cat_id = move.get("category")
            if not cat_id or cat_id not in report:
                continue

            constraints = move.get("start_constraints")
            if constraints and "dir" in constraints and "edge" in constraints:
                variant = f"{constraints['dir']}{constraints['edge']}"
                if variant in required:
                    if variant not in report[cat_id]["implemented"]:
                        report[cat_id]["implemented"].append(variant)
                    if variant in report[cat_id]["missing"]:
                        report[cat_id]["missing"].remove(variant)
            else:
                report[cat_id]["generic_count"] += 1

        return report

File: src/services/engine.ts
import rawConfig from '../../config/moves_config.yaml';
import {
  State,
  ALL_STATES,
  getRelativeConditions,
  getNaturalCurvature,
  calculateNextState,
} from '../domain/models';

export interface MoveRawData {
  id: string;
  name: string;
  category: string;
  difficulty: number;
  turn_rotation?: string;
  conditions: {
    same_foot: boolean;
    same_dir: boolean;
    same_edge: boolean;
  };
  start_constraints?: {
    dir?: string;
    edge?: string;
  };
  geometry_config?: {
    radius_factor?: number;
    sweep_angle_factor?: number;
  };
}

export interface ConfigData {
  categories?: Record<string, string>;
  moves?: MoveRawData[];
}

export interface Move {
  id: string;
  name: string;
  category: string;
  difficulty: number;
  turn_rotation?: string;
  conditions: {
    same_foot: boolean;
    same_dir: boolean;
    same_edge: boolean;
  };
  start_constraints?: {
    dir?: string;
    edge?: string;
  };
  rotation_dir?: 'CW' | 'CCW' | null;
  geometry_config?: {
    radius_factor?: number;
    sweep_angle_factor?: number;
  };
}

export interface MoveOption {
  target_state: State;
  move: Move;
}

export interface TransitionDetail {
  from_state: State;
  to_state: State;
  candidate_moves: Move[];
  selected_move: Move;
}

export interface VerificationResponse {
  valid: boolean;
  error?: string;
  states?: State[];
  transitions?: TransitionDetail[];
  total_difficulty: number;
  is_ambiguous: boolean;
}

export interface MoveVerificationDetail {
  from_state: State;
  move: Move;
  to_state: State;
}

export interface MoveVerificationResponse {
  valid: boolean;
  error?: string;
  trace?: MoveVerificationDetail[];
  total_difficulty: number;
}

export function checkMatch(
  currentState: State,
  targetState: State,
  moveConfig: MoveRawData
): boolean {
  const conditions = moveConfig.conditions || {};
  const actualConditions = getRelativeConditions(currentState, targetState);

  if (
    conditions.same_foot !== actualConditions.same_foot ||
    conditions.same_dir !== actualConditions.same_dir ||
    conditions.same_edge !== actualConditions.same_edge
  ) {
    return false;
  }

  if (moveConfig.start_constraints) {
    const constraints = moveConfig.start_constraints;
    if (constraints.dir && currentState.direction !== constraints.dir) {
      return false;
    }
    if (constraints.edge && currentState.edge !== constraints.edge) {
      return false;
    }
  }

  return true;
}

interface ChoreoSearchNode {
  f_score: number;
  current_state: State;
  path_depth: number;
  path: Array<[State, Move | null]>;
  accumulated_difficulty: number;
  categories_used: Set<string>;
  cw_count: number;
  ccw_count: number;
}

class MinHeap<T> {
  private heap: Array<{ priority: number; id: number; data: T }> = [];

  push(priority: number, id: number, data: T) {
    this.heap.push({ priority, id, data });
    this.bubbleUp(this.heap.length - 1);
  }

  pop(): { priority: number; id: number; data: T } | undefined {
    if (this.heap.length === 0) return undefined;
    const top = this.heap[0];
    const bottom = this.heap.pop()!;
    if (this.heap.length > 0) {
      this.heap[0] = bottom;
      this.sinkDown(0);
    }
    return top;
  }

  get length(): number {
    return this.heap.length;
  }

  private bubbleUp(index: number) {
    while (index > 0) {
      const parent = Math.floor((index - 1) / 2);
      if (this.compare(index, parent) < 0) {
        this.swap(index, parent);
        index = parent;
      } else {
        break;
      }
    }
  }

  private sinkDown(index: number) {
    const len = this.heap.length;
    while (true) {
      let smallest = index;
      const left = 2 * index + 1;
      const right = 2 * index + 2;

      if (left < len && this.compare(left, smallest) < 0) {
        smallest = left;
      }
      if (right < len && this.compare(right, smallest) < 0) {
        smallest = right;
      }

      if (smallest !== index) {
        this.swap(index, smallest);
        index = smallest;
      } else {
        break;
      }
    }
  }

  private compare(i: number, j: number): number {
    if (this.heap[i].priority !== this.heap[j].priority) {
      return this.heap[i].priority - this.heap[j].priority;
    }
    return this.heap[i].id - this.heap[j].id;
  }

  private swap(i: number, j: number) {
    const tmp = this.heap[i];
    this.heap[i] = this.heap[j];
    this.heap[j] = tmp;
  }
}

export class ChoreographyEngine {
  readonly moves: MoveRawData[];
  readonly categories: Record<string, string>;
  readonly distance_matrix: Record<string, Record<string, number>>;

  constructor(configData?: ConfigData) {
    const config = configData || (rawConfig as ConfigData);
    this.moves = config.moves || [];
    this.categories = config.categories || {};
    this.distance_matrix = this.buildDistanceMatrix();
  }

  private buildDistanceMatrix(): Record<string, Record<string, number>> {
    const matrix: Record<string, Record<string, number>> = {};

    for (const start of ALL_STATES) {
      const startStr = start.toString();
      matrix[startStr] = {};
      for (const target of ALL_STATES) {
        matrix[startStr][target.toString()] = Infinity;
      }
    }

    for (const start of ALL_STATES) {
      const startStr = start.toString();
      const queue: Array<[State, number]> = [[start, 0]];
      matrix[startStr][startStr] = 0;
      const visited = new Set<string>([startStr]);

      while (queue.length > 0) {
        const [curr, dist] = queue.shift()!;
        const options = this.getPossibleTransitions(curr);
        for (const opt of options) {
          const nxtStr = opt.target_state.toString();
          if (!visited.has(nxtStr)) {
            visited.add(nxtStr);
            matrix[startStr][nxtStr] = dist + 1;
            queue.push([opt.target_state, dist + 1]);
          }
        }
      }
    }

    return matrix;
  }

  buildMove(moveData: MoveRawData, currentState: State): Move {
    const turnRot = moveData.turn_rotation;
    let absRot: 'CW' | 'CCW' | null = null;

    if (turnRot === 'natural') {
      absRot = getNaturalCurvature(currentState);
    } else if (turnRot === 'opposite') {
      const startCurv = getNaturalCurvature(currentState);
      absRot = startCurv === 'CCW' ? 'CW' : 'CCW';
    }

    return {
      id: moveData.id,
      name: moveData.name,
      category: moveData.category,
      difficulty: moveData.difficulty,
      turn_rotation: moveData.turn_rotation,
      conditions: moveData.conditions,
      start_constraints: moveData.start_constraints,
      rotation_dir: absRot,
      geometry_config: moveData.geometry_config,
    };
  }

  getPossibleTransitions(
    currentState: State,
    maxDifficulty: number = 999
  ): MoveOption[] {
    const results: MoveOption[] = [];

    for (const targetState of ALL_STATES) {
      if (targetState.equals(currentState)) {
        continue;
      }

      for (const moveData of this.moves) {
        if (checkMatch(currentState, targetState, moveData)) {
          const diff = moveData.difficulty ?? 0;
          if (diff <= maxDifficulty) {
            const moveObj = this.buildMove(moveData, currentState);
            results.push({
              target_state: targetState,
              move: moveObj,
            });
          }
        }
      }
    }

    results.sort((a, b) => {
      if (a.move.difficulty !== b.move.difficulty) {
        return a.move.difficulty - b.move.difficulty;
      }
      return a.move.name.localeCompare(b.move.name);
    });

    return results;
  }

  verifySequence(sequenceStr: string): VerificationResponse {
    const parts = sequenceStr
      .split('->')
      .map((p) => p.trim().toUpperCase())
      .filter((p) => p.length > 0);

    const states: State[] = [];
    for (const part of parts) {
      try {
        states.push(State.fromString(part));
      } catch (e: any) {
        return {
          valid: false,
          error: `状态字符 '${part}' 格式有误: ${e.message}`,
          total_difficulty: 0,
          is_ambiguous: false,
        };
      }
    }

    if (states.length < 2) {
      return {
        valid: false,
        error: '状态序列中至少需要包含 2 个有效状态才能进行转移校验。',
        total_difficulty: 0,
        is_ambiguous: false,
      };
    }

    const transitionsDetails: TransitionDetail[] = [];
    let totalDifficulty = 0;

    for (let i = 0; i < states.length - 1; i++) {
      const sFrom = states[i];
      const sTo = states[i + 1];

      if (sFrom.equals(sTo)) {
        return {
          valid: false,
          error: `第 ${i + 1} 步转移出现原地停滞 (${sFrom} -> ${sTo})，这不符合动力学步法转移规则。`,
          total_difficulty: 0,
          is_ambiguous: false,
        };
      }

      const matchedMoves: Move[] = [];
      for (const moveData of this.moves) {
        if (checkMatch(sFrom, sTo, moveData)) {
          matchedMoves.push(this.buildMove(moveData, sFrom));
        }
      }

      if (matchedMoves.length === 0) {
        return {
          valid: false,
          error: `无法识别的物理转移: 从状态 ${sFrom} 无法直接通过任何已知动作转移到 ${sTo}。`,
          total_difficulty: 0,
          is_ambiguous: false,
        };
      }

      matchedMoves.sort((a, b) => {
        if (a.difficulty !== b.difficulty) {
          return a.difficulty - b.difficulty;
        }
        return a.name.localeCompare(b.name);
      });

      transitionsDetails.push({
        from_state: sFrom,
        to_state: sTo,
        candidate_moves: matchedMoves,
        selected_move: matchedMoves[0],
      });
      totalDifficulty += matchedMoves[0].difficulty;
    }

    const isAmbiguous = transitionsDetails.some(
      (t) => t.candidate_moves.length > 1
    );

    return {
      valid: true,
      states,
      transitions: transitionsDetails,
      total_difficulty: totalDifficulty,
      is_ambiguous: isAmbiguous,
    };
  }

  verifyMoveSequence(
    moveIds: string[],
    startState?: State
  ): MoveVerificationResponse {
    if (!moveIds || moveIds.length === 0) {
      return { valid: false, error: '动作 ID 序列不能为空。', total_difficulty: 0 };
    }

    const moveDb: Record<string, MoveRawData> = {};
    for (const m of this.moves) {
      moveDb[m.id] = m;
    }

    let currentState = startState;
    if (!currentState) {
      const firstMoveData = moveDb[moveIds[0]];
      if (!firstMoveData) {
        return {
          valid: false,
          error: `无法识别序列起始处的动作 ID: ${moveIds[0]}`,
          total_difficulty: 0,
        };
      }
      const constraints = firstMoveData.start_constraints || {};
      const dirC = constraints.dir || 'F';
      const edgeC = constraints.edge || 'O';
      currentState = new State('L', dirC, edgeC);
    }

    const traceDetails: MoveVerificationDetail[] = [];
    let totalDifficulty = 0;

    for (let idx = 0; idx < moveIds.length; idx++) {
      const moveId = moveIds[idx];
      const moveData = moveDb[moveId];
      if (!moveData) {
        return {
          valid: false,
          error: `在第 ${idx + 1} 步检测到未知动作 ID: '${moveId}'`,
          total_difficulty: 0,
        };
      }

      const constraints = moveData.start_constraints;
      if (constraints) {
        if (constraints.dir && currentState.direction !== constraints.dir) {
          return {
            valid: false,
            error: `第 ${idx + 1} 步动作校验失败：动作 '${moveData.name}' 要求以 '${constraints.dir}' 向起滑，但当前滑行状态为 '${currentState}'。`,
            total_difficulty: 0,
          };
        }
        if (constraints.edge && currentState.edge !== constraints.edge) {
          return {
            valid: false,
            error: `第 ${idx + 1} 步动作校验失败：动作 '${moveData.name}' 要求以 '${constraints.edge}' 内外刃起滑，但当前滑行状态为 '${currentState}'。`,
            total_difficulty: 0,
          };
        }
      }

      const nextState = calculateNextState(currentState, moveData.conditions);
      const moveObj = this.buildMove(moveData, currentState);

      traceDetails.push({
        from_state: currentState,
        move: moveObj,
        to_state: nextState,
      });

      totalDifficulty += moveObj.difficulty;
      currentState = nextState;
    }

    return {
      valid: true,
      trace: traceDetails,
      total_difficulty: totalDifficulty,
    };
  }

  generateSequence(
    steps: number,
    maxDifficulty: number,
    startState?: State
  ): Array<[State, Move | null]> | null {
    if (steps <= 0) return [];

    const initState =
      startState || ALL_STATES[Math.floor(Math.random() * ALL_STATES.length)];

    const shuffle = <T>(arr: T[]): T[] => {
      const copy = [...arr];
      for (let i = copy.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [copy[i], copy[j]] = [copy[j], copy[i]];
      }
      return copy;
    };

    const dfs = (
      currState: State,
      remainingSteps: number
    ): Array<[State, Move]> | null => {
      if (remainingSteps === 0) return [];

      const options = this.getPossibleTransitions(currState, maxDifficulty);
      if (options.length === 0) return null;

      const shuffledOptions = shuffle(options);

      for (const opt of shuffledOptions) {
        const subPath = dfs(opt.target_state, remainingSteps - 1);
        if (subPath !== null) {
          return [[opt.target_state, opt.move], ...subPath];
        }
      }

      return null;
    };

    const subPath = dfs(initState, steps);
    if (subPath === null) return null;

    const result: Array<[State, Move | null]> = [];
    let curr = initState;
    for (const [nxtState, move] of subPath) {
      result.push([curr, move]);
      curr = nxtState;
    }
    result.push([curr, null]);

    return result;
  }

  checkLibraryIntegrity(): Record<
    string,
    {
      name: string;
      implemented: string[];
      missing: string[];
      generic_count: number;
    }
  > {
    const categories =
      Object.keys(this.categories).length > 0
        ? this.categories
        : {
            three_turn: '转三步 (Three-Turn)',
            bracket: '括弧步 (Bracket)',
            rocker: '摇滚步 (Rocker)',
            counter: '计数步 (Counter)',
            mohawk: '莫霍克步 (Mohawk)',
            choctaw: '乔克陶步 (Choctaw)',
          };

    const required = ['FO', 'FI', 'BO', 'BI'];
    const report: Record<
      string,
      {
        name: string;
        implemented: string[];
        missing: string[];
        generic_count: number;
      }
    > = {};

    for (const [catId, catName] of Object.entries(categories)) {
      report[catId] = {
        name: catName,
        implemented: [],
        missing: [...required],
        generic_count: 0,
      };
    }

    for (const move of this.moves) {
      const catId = move.category;
      if (!catId || !report[catId]) continue;

      const constraints = move.start_constraints;
      if (constraints && constraints.dir && constraints.edge) {
        const variant = `${constraints.dir}${constraints.edge}`;
        if (required.includes(variant)) {
          if (!report[catId].implemented.includes(variant)) {
            report[catId].implemented.push(variant);
          }
          const missIdx = report[catId].missing.indexOf(variant);
          if (missIdx !== -1) {
            report[catId].missing.splice(missIdx, 1);
          }
        }
      } else {
        report[catId].generic_count += 1;
      }
    }

    return report;
  }

  searchPaths(
    startState: State,
    endState: State,
    intermediateCount: number,
    maxDifficulty: number = 5,
    maxResults: number = 10,
    weights?: Record<string, number>
  ): Array<Array<[State, Move | null]>> {
    const targetSteps = intermediateCount + 1;
    const results: Array<Array<[State, Move | null]>> = [];

    const w = weights || {};
    const C_STEP = w.step_cost ?? 10.0;
    const C_DIFF = w.difficulty_bonus ?? 3.0;
    const C_BALANCE = w.balance_penalty ?? 15.0;
    const C_DIVERSITY = w.diversity_penalty ?? 20.0;

    const TARGET_CATEGORIES = 4;

    const openSet = new MinHeap<ChoreoSearchNode>();
    let nodeId = 0;

    const initNode: ChoreoSearchNode = {
      f_score: 0.0,
      current_state: startState,
      path_depth: 0,
      path: [[startState, null]],
      accumulated_difficulty: 0,
      categories_used: new Set<string>(),
      cw_count: 0,
      ccw_count: 0,
    };

    openSet.push(initNode.f_score, nodeId, initNode);
    const closedSet = new Set<string>();

    while (openSet.length > 0 && results.length < maxResults) {
      const popped = openSet.pop();
      if (!popped) break;
      const curr = popped.data;

      if (curr.path_depth === targetSteps) {
        if (curr.current_state.equals(endState)) {
          results.push(curr.path);
        }
        continue;
      }

      const catSorted = Array.from(curr.categories_used).sort().join(',');
      const stateKey = `${curr.current_state.toString()}_${curr.path_depth}_${catSorted}_${curr.cw_count}_${curr.ccw_count}`;

      if (closedSet.has(stateKey)) {
        continue;
      }
      closedSet.add(stateKey);

      const options = this.getPossibleTransitions(
        curr.current_state,
        maxDifficulty
      );

      for (const opt of options) {
        const nextState = opt.target_state;
        const move = opt.move;

        const remainingSteps = targetSteps - (curr.path_depth + 1);
        const minDist =
          this.distance_matrix[nextState.toString()]?.[endState.toString()] ??
          Infinity;

        if (remainingSteps < minDist) {
          continue;
        }

        const nextCategories = new Set(curr.categories_used);
        nextCategories.add(move.category);

        const nextCw = curr.cw_count + (move.rotation_dir === 'CW' ? 1 : 0);
        const nextCcw = curr.ccw_count + (move.rotation_dir === 'CCW' ? 1 : 0);

        const nextPath: Array<[State, Move | null]> = [
          ...curr.path.slice(0, -1),
          [curr.current_state, move],
          [nextState, null],
        ];

        const gScore =
          (curr.path_depth + 1) * C_STEP -
          (curr.accumulated_difficulty + move.difficulty) * C_DIFF;

        const hDist = remainingSteps * C_STEP;
        const balDiff = Math.abs(nextCw - nextCcw);
        const minFinalBalDiff = Math.max(0, balDiff - remainingSteps);
        const hBal = minFinalBalDiff * C_BALANCE;

        const missingCats = Math.max(
          0,
          TARGET_CATEGORIES - nextCategories.size - remainingSteps
        );
        const hDiv = missingCats * C_DIVERSITY;

        const fScore = gScore + hDist + hBal + hDiv;

        const nextNode: ChoreoSearchNode = {
          f_score: fScore,
          current_state: nextState,
          path_depth: curr.path_depth + 1,
          path: nextPath,
          accumulated_difficulty: curr.accumulated_difficulty + move.difficulty,
          categories_used: nextCategories,
          cw_count: nextCw,
          ccw_count: nextCcw,
        };

        nodeId += 1;
        openSet.push(fScore, nodeId, nextNode);
      }
    }

    return results;
  }
}